seed/seedwriter/writer.go

```diff
@@ -302,49 +340,118 @@ func (w *Writer) SetOptionsSnaps(optSnaps []*OptionSnap) error {
 				return err
 			}
 		}
+		if local {
+			w.localSnaps = append(w.localSnaps, &SeedSnap{
+				SnapRef:    nil,
+				Path:       sn.Path,
+				local:      true,
+				optionSnap: sn,
+			})
+		}
 	}
 
 	return nil
 }
 
-func (w *Writer) Start(db asserts.RODatabase, newFetcher NewFetcherFunc) error {
+// Start starts the seed writing. It creates a RefAssertsFetcher using
+// newFetcher and uses it to fetch model related assertions. For
+// convenience it returns the fetcher possibly for use to fetch seed
+// snap assertions, a task that the writer delegates as well as snap
+// downloading. The writer assumes that the snap assertions will end up
+// in the given db (writing assertion database).
+func (w *Writer) Start(db asserts.RODatabase, newFetcher NewFetcherFunc) (RefAssertsFetcher, error) {
 	if err := w.checkStep(startStep); err != nil {
-		return err
+		return nil, err
 	}
 	if db == nil {
-		return fmt.Errorf("internal error: Writer *asserts.RODatabsae is nil")
+		return nil, fmt.Errorf("internal error: Writer *asserts.RODatabase is nil")
 
 	}
 	if newFetcher == nil {
-		return fmt.Errorf("internal error: Writer newFetcherFunc is nil")
+		return nil, fmt.Errorf("internal error: Writer newFetcherFunc is nil")
 	}
 	w.db = db
 
 	f := MakeRefAssertsFetcher(newFetcher)
 
 	// XXX support UBUNTU_IMAGE_SKIP_COPY_UNVERIFIED_MODEL ?
 	if err := f.Save(w.model); err != nil {
-		return fmt.Errorf("cannot fetch and check prerequisites for the model assertion: %v", err)
+		return nil, fmt.Errorf("cannot fetch and check prerequisites for the model assertion: %v", err)
 	}
 
 	w.modelRefs = f.Refs()
 
 	// XXX get if needed the store assertion
 
-	return w.tree.mkFixedDirs()
+	if err := w.tree.mkFixedDirs(); err != nil {
+		return nil, err
+	}
+
+	return f, nil
 }
 
-// LocalSnaps()
-// InfoDerived()
+// LocalSnaps returns a list of seed snaps that are local.  The writer
+// delegates to produce *snap.Info for them to then be set via
+// SetInfo. If matching snap assertions can be found as well they can
+// be passed into SeedSnap ARefs, assuming they were added to the
+// writing assertion database.
+func (w *Writer) LocalSnaps() ([]*SeedSnap, error) {
+	if err := w.checkStep(localSnapsStep); err != nil {
+		return nil, err
+	}
+
+	return w.localSnaps, nil
+}
+
+// InfoDerived checks the local snaps metadata provided via setting it
+// into the SeedSnaps returned by the previous LocalSnaps.
+func (w *Writer) InfoDerived() error {
+	if err := w.checkStep(infoDerivedStep); err != nil {
+		return err
+	}
+
+	for _, sn := range w.localSnaps {
+		if sn.Info == nil {
+			return fmt.Errorf("internal error: at this point snap %q Info should have been set", sn.Path)
```

Comment:
Not super important since this is an internal error, but I think it would be good to simply say what is the actual point we are at here, so it's not neccessary to lookup it up in the code to understand the context of the failure.

Reply:
I just noticed this error pattern is used in other places and in the already landed code; if you think it's a sensible suggestion then a followup is fine.

Reply:
yes,  a follow up is probably best